### PR TITLE
Switch to be a subordinate charm deployed against openstack-integrator

### PR DESCRIPTION
**Pull request by addyess**
This charm hasn't ever been published to charmhub.  So upgrades weren't something to be addressed

This helps the `openstack-cloud-controller` follow in-line with other charms like `aws-cloud-provider`, and `vsphere-cloud-providers` which deploy against their "integrator" charm as a subordinate, and use the kubernetes credentials provided via `kube-control` and the api certs provided via `tls-interface` to apply a specific mutated upstream version of `openstack-cloud-controller` into the cluster.

It also exposes the `external-control-plane` relation to k8s-cp

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,6 @@
-ops>=1.3.0,<2.0.0
 lightkube>=0.10.1,<1.0.0
-serialized-data-interface>=0.4.0,<1.0.0
+ops>=1.3.0,<2.0.0
+ops.manifest>=1.1.3,<2.0.0
+pydantic
+git+https://github.com/charmed-kubernetes/interface-kube-control.git@6dd289d1c795fdeda1bed17873b8d6562227c829#subdirectory=ops
+git+https://github.com/juju-solutions/interface-openstack-integration.git@0f81bb414ad8676ab889bbba24a34cffd12ce2ca#subdirectory=ops
```
Comment (addyess) [inline requirements.txt:6]: DO NOT MERGE with this requirement -- ensure it's from charmed-kubernetes main branch commit

File: config.yaml
Comment (mateoflorido) [inline config.yaml:1]: Where did these configurations go? Aren't they necessary now?

File: metadata.yaml [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
-name: openstack-cloud-controller-operator
-display-name: OpenStack Cloud Controller Operator
+name: openstack-cloud-controller
+display-name: OpenStack Cloud Controller
 summary: Runs the OpenStack Cloud Controller in the cluster.
 description: >-
   The OpenStack cloud provider provides the Kubernetes cluster access to
   native resources from OpenStack, such as storage and load balancers.
   This operator includes the core Cloud Controller component.
+docs: https://discourse.charmhub.io/t/openstack-cloud-controller-docs-index/????
```
Comment (mateoflorido) [inline metadata.yaml:10]: Same nit 😄

File: config.yaml
Comment (addyess) [reply config.yaml:1]: yeah, all of these config items comes from the relation to openstack-integrator.  When this was a k8s charm, it was necessary to provide them via config (or a CMR).   If we want to have them use different creds in the future, you could -- but you can configure those changes on the `openstack-integrator` instead.

Comment (mateoflorido) [reply config.yaml:1]: Nice, thanks 👍🏻

**Comment by nobuto-m** [comment]
Oh I didn't realized this change until now. What I liked in the original approach was that it was able to take all info passed through the charm config so it doesn't have to be Charmed Kubernetes always.
https://github.com/canonical/openstack-cloud-controller-operator/pull/4/files#diff-d8d0422389f03d783e32e627250fe29834bd09c6361640d1ff00661dd6820034

One possible use case was that when deploying MicroK8s on top of any OpenStack. By deploying this charm, MicroK8s could provision a load balancer resource from Octavia with Juju, instead of manually installing cloud-provider-openstack and configure it by hand.

**Comment by addyess** [comment]
No problem @nobuto-m .  I'd bet we could still accommodate this approach with the config.  What if we target a feature to relate with microk8s as a subordinate too?

**Comment by nobuto-m** [comment]
> No problem @nobuto-m . I'd bet we could still accommodate this approach with the config. What if we target a feature to relate with microk8s as a subordinate too?

That may be a good idea too. One motivation using microk8s standalone without charms may be that the underlying OpenStack is restrictive for bootstrapping Juju controllers (Juju agent binary index or Juju image index may not be available). Overriding info supposed to be passed through the relation by using config would be a plus for those situations.

**Comment by addyess** [comment]
okay, so--- i understand the need for config override -- etc.  That makes alot of sense.  I also understand the need to deploy with a microk8s worker deployed on an openstack instance.  

I don't quite follow your thoughts on what model this charm stands.  Are you thinking a juju controller bootstrapped in a kubernetes model?  That would mean this charm would need to install into a k8s environment -- is that what you're proposing?

As i understand it there's no current way to make the SAME charm be both a subordinate and a standalone k8s charm.  Maybe we'd have to build two separate charms with effectively the same

**Comment by nobuto-m** [comment]
> I don't quite follow your thoughts on what model this charm stands. Are you thinking a juju controller bootstrapped in a kubernetes model? That would mean this charm would need to install into a k8s environment -- is that what you're proposing?

Yes. I might have mixed up with something else, but when this `openstack-cloud-controller-operator` charm was written initially, I was under the impression that it was going to be the new charm as a k8s charm to be deployed on top of Kubernetes and drop the necessity of deploying the `openstack-integrator` charm and the cdk-addon for the cloud-provider-openstack portion.

But while writing this text, I realized that deploying some charms on top of Kubernetes doesn't necessarily mean the charms can update kube-apiserver extra flags such as `--cloud-provider=external`. That's why relations are necessary to change something on the underlying Kubernetes. Sounds like my assumed user story is half baked in my head.

**Comment by addyess** [comment]
There's an awkward little dance to bootstrap juju on k8s without storage which has kinda pushed us towards this model of making charms of this type as machine subordinates rather than k8s charms: 

Some examples include
* vsphere-cloud-provider (includes storage manifests)
* azure-cloud-provider (includes storage manifests)
* aws-cloud-provider  (storage manifest charm is separate)

Along with these charm are the cloud storage charms which end up being very similar
* aws-k8s-storage
* gcp-k8s-storage

You CAN bootstrap a controller in kubernetes without storage, then use these charms to add cloud-based storage access to the cluster.  It's just not totally straight forward.

For charmed-kubernetes we found making subordinate charms off the integrator charm was much easier at authorizing the charms, (using the trust permissions of the integrator), they could relate to the control-plane to provide the external cloud-provider, and  they don't require bootstrapping a controller ignoring storage in order to install the storage manifests into the cluster.  

But there COULD be some value in still letting folks go through that if they need this charm outside of charmed-kubernetes

**Pull request merged**